Avoid extra `Array.Copy` in `List<T>.Insert(Range)` when the list is full

karakasa opened the pull request:
Originally there are 3 `Array.Copy` during `List<T>.Insert` when `Capacity` is maxed out. Avoid one extra copy with a modified `Grow`.

[Microbenchmark in following comments.](https://github.com/dotnet/runtime/pull/90089#issuecomment-1667807944) `List<T>.InsertRange` also benefits from similar patterns. Performance is ~20% better for best cases.

Fix #89915

karakasa committed:
Avoid extra copy when insertting on `List` at capacity

Originally there are 3 `Array.Copy` during `List<T>.Insert`
when `Capacity` is maxed out. Avoid this extra copy with a modified `Grow`.

Fix #89915

karakasa committed:
typo

karakasa committed:
fix typo

EgorBo commented:
If I understand it correctly, this gives us up to 3.5% imprvoement for the best case - insertion at the 0 index in a large array, right?

karakasa commented:
> If I understand it correctly, this gives us up to 3.5% imprvoement for the best case - insertion at the 0 index in a large array, right?

The gain is 20%-ish for insertion at 0 into 10000-long list. 

Type | Size | Time of Insert@0 (`generation+insertion` minus `generation`-only)
-- | -- | --
Original | 100 | 67.6 ns
Improved | 100 | 45.7 ns
Original | 10000 | 3888.6 ns
Improved | 10000 | 3070.2 ns

The correct benchmark should utilize `[IterationSetup]`. But for the sake of simplicity, I made two benchmarks (one generates the list only, one generates and inserts), and subtract the `generation+insertion` time with `generation`-only time.

karakasa commented:
@EgorBo I think `List<T>.InsertRange` may benefit from similar optimization. Is it appropriate to put it within this PR?

karakasa committed:
Remove impossible branches

karakasa committed:
fixed semicolon typo

karakasa committed:
improve InsertRange perf

karakasa committed:
Remove redundant capacity argument

karakasa committed:
Remove unnecessary assertion

karakasa commented:
btw, should we replace `Array.Copy` with `Span.Copy` in List<T> when possible? Since there's no type-safety issues. All types are `T`.

karakasa commented:
> Maybe.

I just did a microbenchmark and it seems replacing `Array.Copy` with `Span.CopyTo` doesn't have observable impact on perf, even for 10000-long `int` list. So I removed that reply.

|              Method |  Size |        Mean |     Error |    StdDev |
|-------------------- |------ |------------:|----------:|----------:|
|   **ImprovedInsertOne** |   **100** |    **205.8 ns** |   **1.36 ns** |   **1.13 ns** |
|       SpanInsertOne |   100 |    198.0 ns |   2.52 ns |   2.36 ns |
| ImprovedInsertRange |   100 |    215.8 ns |   1.53 ns |   1.35 ns |
|     SpanInsertRange |   100 |    232.6 ns |   0.79 ns |   0.70 ns |
|   **ImprovedInsertOne** | **10000** | **15,929.9 ns** |  **50.44 ns** |  **47.18 ns** |
|       SpanInsertOne | 10000 | 15,896.6 ns |  76.68 ns |  71.73 ns |
| ImprovedInsertRange | 10000 | 16,916.3 ns |  63.25 ns |  52.82 ns |
|     SpanInsertRange | 10000 | 16,902.4 ns | 107.19 ns | 100.27 ns |

> This PR is a similar tradeoff at smaller scale. It makes each `List<T>` type instantiation more expensive to create and take more memory.

Yes you are right.

MichalPetryka commented:
> I just did a microbenchmark and it seems replacing `Array.Copy` with `Span.CopyTo` doesn't have big observable perf impact, even for 10000-long `int` list.

The bigger impact here would be with smaller collections, since `Span.Copy` then skips a few checks and allows unrolling.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -465,6 +465,38 @@ internal void Grow(int capacity)
+
+            int capacity = checked(_size + insertionCount);
+
+            // Follow the same logic from Grow(capacity)
```
We should rather move the common logic to a helper method, reuse it in both `GrowForInsertion` and `Grow` and ensure it's getting linlined in both cases.

```cs
private int GetNewCapacity(int capacity)
{
    Debug.Assert(_items.Length < capacity);

    int newCapacity = _items.Length == 0 ? DefaultCapacity : 2 * _items.Length;

    // Allow the list to grow to maximum possible capacity (~2G elements) before encountering overflow.
    // Note that this check works even when _items.Length overflowed thanks to the (uint) cast
    if ((uint)newCapacity > Array.MaxLength) newCapacity = Array.MaxLength;

    // If the computed capacity is still less than specified, set to the original argument.
    // Capacities exceeding Array.MaxLength will be surfaced as OutOfMemoryException by Array.Resize.
    if (newCapacity < capacity) newCapacity = capacity;
    
    return newCapacity;
} 
```

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -465,6 +465,38 @@ internal void Grow(int capacity)
+
+            T[] newItems = new T[newCapacity];
+            if (indexToInsert != 0)
+                Array.Copy(_items, newItems, indexToInsert);
```
nit: it would be nice to name the last argument to make it clear that it's length, not index (and a bug)

```suggestion
                Array.Copy(_items, newItems, length: indexToInsert);
```

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -465,6 +465,38 @@ internal void Grow(int capacity)
+            if (indexToInsert != 0)
+                Array.Copy(_items, newItems, indexToInsert);
+
+            if (_size != indexToInsert)
+                Array.Copy(_items, indexToInsert, newItems, indexToInsert + insertionCount, _size - indexToInsert);
```
nit: please add the brackets (I know the style in this file is not consistent)

```suggestion
            if (_size != indexToInsert)
            {
                Array.Copy(_items, indexToInsert, newItems, indexToInsert + insertionCount, _size - indexToInsert);
            }
```

adamsitnik reviewed (changes requested):
@karajas Thank you for your contribution!

PTAL at my comments, apply the changes and share updated perf numbers. Thanks!

karakasa committed:
Apply suggestions from code review

code formatting

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

karakasa committed:
Put shared logic into a standalone method

Shared logics of getting new capacities in
Grow & GrowForInsertion are moved to
a separated method.

karakasa commented:
> share updated perf numbers.

| Method         | Toolchain | Size  | Mean (ns) | Ratio |
|----------------|-----------|------:|----------:|------:|
| Add            | 8 RC2     | 100   | 51.40     | 1.00  |
| Add            | PR        | 100   | 51.30     | 1.00  |
| Add            | PR-Reuse  | 100   | 50.40     | 0.98  |
| AddRange       | 8 RC2     | 100   | 63.70     | 1.00  |
| AddRange       | PR        | 100   | 61.70     | 0.97  |
| AddRange       | PR-Reuse  | 100   | 63.60     | 1.00  |
| InsertAt0      | 8 RC2     | 100   | 46.80     | 1.00  |
| InsertAt0      | PR        | 100   | 42.60     | 0.91  |
| InsertAt0      | PR-Reuse  | 100   | 41.00     | 0.88  |
| InsertRangeAt0 | 8 RC2     | 100   | 70.70     | 1.00  |
| InsertRangeAt0 | PR        | 100   | 61.00     | 0.86  |
| InsertRangeAt0 | PR-Reuse  | 100   | 62.50     | 0.88  |
| Add            | 8 RC2     | 10000 | 3,655.30  | 1.00  |
| Add            | PR        | 10000 | 3,452.70  | 0.94  |
| Add            | PR-Reuse  | 10000 | 3,478.50  | 0.95  |
| AddRange       | 8 RC2     | 10000 | 3,849.00  | 1.00  |
| AddRange       | PR        | 10000 | 3,695.20  | 0.96  |
| AddRange       | PR-Reuse  | 10000 | 3,631.80  | 0.94  |
| InsertAt0      | 8 RC2     | 10000 | 3,717.20  | 1.00  |
| InsertAt0      | PR        | 10000 | 2,674.50  | 0.72  |
| InsertAt0      | PR-Reuse  | 10000 | 2,654.10  | 0.71  |
| InsertRangeAt0 | 8 RC2     | 10000 | 4,908.90  | 1.00  |
| InsertRangeAt0 | PR        | 10000 | 3,610.20  | 0.74  |
| InsertRangeAt0 | PR-Reuse  | 10000 | 3,666.20  | 0.75  |

<details><summary>code</summary>

````
public class InsertIntoExisting
{
    [Params(100, 10000)]
    public int Size { get; set; }
    private int[] _array;
    [GlobalSetup]
    public void InitArrayForInsertRange()
    {
        _array = new int[Size];
    }
    public List<int> PrepareList()
    {
        var list = new List<int>(Size);
        var size = Size;
        for (var i = 0; i < size; i++)
            list.Add(i);

        return list;
    }

    [Benchmark]
    public object Baseline()
    {
        return PrepareList();
    }
    [Benchmark]
    public object InsertAt0()
    {
        var list = PrepareList();
        list.Insert(0, 0);
        return list;
    }
    [Benchmark]
    public object InsertRangeAt0()
    {
        var list = PrepareList();
        list.InsertRange(0, _array);
        return list;
    }
    [Benchmark]
    public object Add()
    {
        var list = PrepareList();
        list.Add(0);
        return list;
    }
    [Benchmark]
    public object AddRange()
    {
        var list = PrepareList();
        list.AddRange(_array);
        return list;
    }
}
````
</details>

| Method           | Toolchain | Size | Mean        |
|------------------|-----------|-----:|------------:|
| Add_OneByOne     | 8 RC2     | 8    | 34.48 ns    |
| Add_OneByOne     | PR        | 8    | 32.99 ns    |
| Add_OneByOne     | PR-Reuse  | 8    | 32.49 ns    |
| Prepend_OneByOne | 8 RC2     | 8    | 90.57 ns    |
| Prepend_OneByOne | PR        | 8    | 87.40 ns    |
| Prepend_OneByOne | PR-Reuse  | 8    | 88.22 ns    |
| Add_OneByOne     | 8 RC2     | 32   | 96.94 ns    |
| Add_OneByOne     | PR        | 32   | 98.04 ns    |
| Add_OneByOne     | PR-Reuse  | 32   | 97.90 ns    |
| Prepend_OneByOne | 8 RC2     | 32   | 433.29 ns   |
| Prepend_OneByOne | PR        | 32   | 418.13 ns   |
| Prepend_OneByOne | PR-Reuse  | 32   | 412.85 ns   |


<details><summary>code</summary>

````
public class OneByOneTest
{
    [Params(8, 32)]
    public int Size { get; set; }
    [Benchmark]
    public object Prepend_OneByOne()
    {
        var list = new List<int>();
        var size = Size;

        for (var i = 0; i < size; i++)
            list.Insert(0, 0);

        return list;
    }
    [Benchmark]
    public object Add_OneByOne()
    {
        var list = new List<int>();
        var size = Size;

        for (var i = 0; i < size; i++)
            list.Add(0);

        return list;
    }
}
````
</details>

adamsitnik reviewed (approved):
LGTM, thank you for your contribution @karakasa !

karakasa commented:
@adamsitnik btw do you have thoughts on how to benchmark this, which requires a new `List<T>` for every iteration? Obviously I cannot use `[IterationSetup]` because each iteration is too short (and cannot be extended through loop)

Currently I added a baseline function and subtract results, but it can be unstable and troublesome.

```
    [Benchmark]
    public object Baseline()
    {
        return PrepareList();
    }

    [Benchmark]
    public object InsertAt0()
    {
        var list = PrepareList();
        list.Insert(0, 0);
        return list;
    }
```

adamsitnik commented:
@karakasa this is tricky. You can combine `[IterationSetup]` with `[InvocationCount]` which tells how many times a benchmark should be invoked per iteration and then in every benchmark ivocation, access an n-th List that is pre-configured in the setup. We do that for sorting in dotnet/performance:

https://github.com/dotnet/performance/blob/bc0b657f6a759c4a07589a6b5d66818345813304/src/benchmarks/micro/libraries/System.Collections/Sort.cs#L19

https://github.com/dotnet/performance/blob/bc0b657f6a759c4a07589a6b5d66818345813304/src/benchmarks/micro/libraries/System.Collections/Sort.cs#L37-L45

stephentoub commented:
It doesn't necessarily require a new List. For a small amount of overhead, it could do:

CollectionsMarshal.SetCount(list, 0);
list.Capacity = 0;

glenn-slayden commented:
Original issue: https://github.com/dotnet/runtime/issues/89915 (by me)
